torch/_higher_order_ops/cond.py

```diff
@@ -329,10 +330,13 @@ def _same_meta_except_requires_grad(true_out, false_out):
 def cond_op_dense(pred, true_fn, false_fn, operands):
     mode = _get_current_dispatch_mode()
     assert mode is None, "Mode should never be enabled for CPU/CUDA key"
-    if pred:
-        return true_fn(*operands)
+    if not torch.cuda.is_current_stream_capturing():
+        if pred:
+            return true_fn(*operands)
+        else:
+            return false_fn(*operands)
```

Comment:
Yikes, this is a very important point that testing so far has not captured. Maybe the proper approach is to use a new dispatch key, but I'm not sure how this would work for eager mode, since we don't have an obvious way to "undo" side effects. Of course, the side effects are limited. But they are still there and not really possible to reason about. Let me think a little bit about this.

Reply:
So I've been looking into this today. Thank you for pointing it out, because I totally forgot about it.

I do think it needs to be addressed in the PR. Right now, the following will fail in a fresh process:

```
import torch

torch.set_default_device("cuda")

def f(pred, x, filters):
    return torch.cond(pred, lambda y: torch.sum(y), lambda y: torch.sum(torch.nn.functional.conv1d(y, filters)), [x])

g = torch.cuda.CUDAGraph()

pred = torch.tensor(True)
x = torch.randn(33, 16, 30)
filters = torch.randn(20, 16, 5)

f(pred, x, filters)
# f(not pred, x, filters) 

with torch.cuda.graph(g, capture_error_mode="thread_local"):
    f(pred, x, filters)
```

The failure looks like this:

```
$ python repros/stream_capture_cond_cudnn.py
Traceback (most recent call last):
  File "/home/dgalvez/scratch/code/asr/pytorch/torch/fx/graph_module.py", line 303, in __call__
    return super(self.cls, obj).__call__(*args, **kwargs)  # type: ignore[misc]
  File "/home/dgalvez/scratch/code/asr/pytorch/torch/nn/modules/module.py", line 1716, in _wrapped_call_impl
    return self._call_impl(*args, **kwargs)
  File "/home/dgalvez/scratch/code/asr/pytorch/torch/nn/modules/module.py", line 1727, in _call_impl
    return forward_call(*args, **kwargs)
  File "<eval_with_key>.1", line 6, in forward
    conv1d = torch.conv1d(l_args_3_0__1, l_args_2_closure_0_cell_contents_false_branch);  l_args_3_0__1 = l_args_2_closure_0_cell_contents_false_branch = None
RuntimeError: CUDA error: CUBLAS_STATUS_NOT_INITIALIZED when calling `cublasCreate(handle)`

Call using an FX-traced Module, line 6 of the traced Module's generated forward function:
    l_args_3_0__1 = l_args_3_0_
    conv1d = torch.conv1d(l_args_3_0__1, l_args_2_closure_0_cell_contents_false_branch);  l_args_3_0__1 = l_args_2_closure_0_cell_contents_false_branch = None

~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ <--- HERE
    sum_1 = torch.sum(conv1d);  conv1d = None

    return (sum_1,)

Segmentation fault
```

Basically, conv1d depends upon cublas having been initialized.

At the same time, the program could enter a corrupt state if you simply tried to warmup every path by running both sides. This is particularly true in the case of training, which I know isn't supported yet for torch.cond, or otherwise any situation where a buffer is updated (consider batch norm, for example, where the mean and variance are updated by default even when self.training is set to false).

I think that the right approach is to add a new TorchDispatchMode, called ControlFlowWarmupDispatchMode that will always enter "relaxed" stream capture mode upon entry in a new stream. It will run most ops normally, except that it will run both sides of torch.cond, and always run the body of torch.while_loop. That graph will then be discarded, since the whole purpose was to make sure that all ops have been warmed up. This doesn't work for ops with data-dependent output shapes, but we are in luck here: torch.cond and torch.while_loop require static shape analysis to guarantee that the output shapes are always the same, so we know that those won't be used.

I need to make unit tests for this that always run in a subprocess, which I see how to do with run_test.py. And there might be things I am missing here. It will probably take a little bit of time to straighten this out.